### PR TITLE
Make DebugSession.name writable; fixes #79583

### LINKED ISSUE #79583: Debugger: when inlining thread into session, use thread's name as a label
When the web page navigates between the URLs, its thread name changes. I'd be great to either allow session to be renamed or to render the thread name in the tree instead of the session name. I can force thread rename via removing and re-adding it back today, but I can't force the session rename.

Forked from https://github.com/microsoft/vscode/issues/79121

@isidorn , @weinand

### PR DESCRIPTION
**Pull request by dgozman**
@isidorn I took a stub on issue #79583, could you please take a look? Do you think this is a right direction?

I thought about constraining the name setter only to the extension which provides the debugger type for this DebugSession, but wasn't able to get a hold on `extension` to implement this check. Any thoughts?

This is changing extension API, so perhaps [this process](https://github.com/Microsoft/vscode/wiki/Extension-API-process) applies? I am not sure how I can follow up there.

Here is a [branch](https://github.com/dgozman/vscode-mock-debug/tree/debug-session-set-name) I used for testing, it defines "Rename active session" command.

### CODE REVIEW

**Comment by isidorn** [comment]
@dgozman thanks for this PR, I think this direction makes sense but I will also consult with @weinand tomorrow. 

As for the API, this would be a minor API update, so I think we can keep the process about adding new API to the minimum and I can take care of that.

Regarding the actuall code changes I have left comments in the code directly. Thanks a lot!

File: src/vs/workbench/api/node/extHostDebugService.ts
Comment (isidorn) [inline src/vs/workbench/api/node/extHostDebugService.ts]: I do not like that we also change the `this._configuration.name`
More about this in a comment below

File: src/vs/workbench/contrib/debug/browser/debugSession.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/debugSession.ts]: I do not think we should change the underlying `configuration` object.
This to me feels like we should introduce a new attribute on the session called `name`.
And in the `getLabel` we use the `name` of the session, if the `name` is not defined then go back to providing the `configuration.name` as a fallback.

File: src/vs/workbench/contrib/debug/browser/loadedScriptsView.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/loadedScriptsView.ts]: Changes in the `loadedScriptsView` I leave up to @weinand to review.

**Review by isidorn — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/debug/browser/debugSession.ts
Comment (isidorn) [reply src/vs/workbench/contrib/debug/browser/debugSession.ts]: That way we have two independt attributes, and we do not mix concepts between the name of the configuration and the name of the session.
Let me know what you think.

File: src/vs/workbench/api/node/extHostDebugService.ts
Comment (weinand) [reply src/vs/workbench/api/node/extHostDebugService.ts]: I agree: `configuration.name` is used as the initial name for the session but becomes decoupled from the session's name as soon as the session's name is changed.

File: src/vs/workbench/contrib/debug/browser/debugSession.ts
Comment (weinand) [reply src/vs/workbench/contrib/debug/browser/debugSession.ts]: Yes, that's the approach we should use.

**Comment by isidorn** [comment]
For the API I have created this feature request https://github.com/microsoft/vscode/issues/80257
We will discuss this in todays API sync meeting and I will provide the update here.

File: src/vs/vscode.d.ts
Comment (weinand) [inline src/vs/vscode.d.ts]: "The debug session's name is initially taken from the [debug configuration](#DebugConfiguration). Any changes will be properly reflected in the UI."

File: src/vs/workbench/api/node/extHostDebugService.ts
Comment (weinand) [reply src/vs/workbench/api/node/extHostDebugService.ts]: yes, don't touch `this._configuration.name`

**Review by weinand — COMMENTED** [review]
What is missing from the PR is the change propagation from VS Code into an extension. Use case: one extension updates a session's name. This flows nicely back into VS Code and updates the UI but the name change is not propagated to debug session copies that might live in other extensions.

To make this API complete, we would actually need listener API for tracking the name changes of debug sessions.

File: src/vs/workbench/contrib/debug/browser/loadedScriptsView.ts
Comment (weinand) [reply src/vs/workbench/contrib/debug/browser/loadedScriptsView.ts]: looks good to me...

**Comment by isidorn** [comment]
After discussing this in the API sync we agreed:
* This approach is good.
* We can add the name change event later if it is needed
* We should not worry about other extensions being able to change the name (there are worse things since they can send customRequests already)

@dgozman so it would be great if you tackle my comments in the code and we can merge this in. Thanks a lot

**Comment by dgozman** [comment]
Thank you for looking at this, I will address all comments. One question on the last comment.

> What is missing from the PR is the change propagation from VS Code into an extension. Use case: one extension updates a session's name. This flows nicely back into VS Code and updates the UI but the name change is not propagated to debug session copies that might live in other extensions.

I wanted to do that, but wasn't sure about objects relationship. It seems that MainThreadDebugService has a single `_proxy: ExtHostDebugServiceShape` to notify about changes, so I didn't see a way to broadcast to multiple ones. Now after your comment, I suspect there are multiple MainThreadDebugService instances so that each of them should update?

File: src/vs/vscode.d.ts
Comment (dgozman) [reply src/vs/vscode.d.ts]: Updated.

File: src/vs/workbench/api/node/extHostDebugService.ts
Comment (dgozman) [reply src/vs/workbench/api/node/extHostDebugService.ts]: Done.

File: src/vs/workbench/contrib/debug/browser/debugSession.ts
Comment (dgozman) [reply src/vs/workbench/contrib/debug/browser/debugSession.ts]: Sounds good, updated.

**Comment by isidorn** [comment]
@dgozman after discussions we decided to not expose the change event. So that will not be needed. Thanks a lot.

**Review by dgozman — COMMENTED** [review]
Please take another look.

File: src/vs/workbench/api/browser/mainThreadDebugService.ts
Comment (isidorn) [inline src/vs/workbench/api/browser/mainThreadDebugService.ts]: No need to pass `true` here.
Inactive sessions are dead sessions which are still only shown in the Debug Console. I think we should not support renaming them.

File: src/vs/workbench/contrib/debug/browser/callStackView.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/callStackView.ts]: There is no need to refresh the whole tree when the label of the session changes.
it is enough just to re-render the debug session element.
So I would go for `this.tree.rerender(s)` so the change in the view is minimal

File: src/vs/workbench/contrib/debug/browser/debugActionViewItems.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/debugActionViewItems.ts]: Please use `this.getSessions` instead.
We have that helper method since we have two action items and each displays different sessions. The one in the debug console displays inactive sessions, and the one in toolbar not (and it might dislpay subsessions depending on a setting)

File: src/vs/workbench/contrib/debug/browser/debugSession.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/debugSession.ts]: I do not understand why do we need `onDidChangeName`?
Isn't `onDidChangeLabel` enough? Or the other way around - point being I think we can survive with only one.

**Review by isidorn — CHANGES_REQUESTED** [review]

**Comment by isidorn** [comment]
@dgozman thanks a lot for addressing my initial feedback.
I have done an additional review, can you check out my comments and try to adress those if they make sense to you?

Great for the testing branch. Once you wrap up the changes can you test everything again just so we are sure all is good. I will also test it before merging. Things to keep an eye on when testing:
* Call Stack View gets properly updated
* Debug Toolbar dropdown gets updated
* Debug Console session dropdown is also up to date
* Other extensions when accessing the debug sessions always get the name which is up to date. So a great test would be to have one extensions which changes the name every `n` seconds, and another extensions gets the session and writes the name. Just so we are sure the data nicely flows through

**Review by dgozman — COMMENTED** [review]
@isidorn Thanks, I have addressed the feedback. All the cases you mentioned work for me locally, including the two extensions case. I also checked the loaded sources view, and it does update as well.

**Comment by isidorn** [comment]
This works great, I have tested this out and I have found only one issue which I have captured here https://github.com/microsoft/vscode/issues/80379

Thanks a lot for this PR, nice work!

**Comment by dgozman** [comment]
Thank you, I will look at #80379.

**Pull request merged**
